MOBT-2-Swift-Foundation.playground/Contents.swift

```diff
@@ -0,0 +1,332 @@
+            if error != nil { return }
+            guard let data = data else { return }
+            do {
+                let results = try JSONDecoder().decode(Currency.self, from: data)
+                self.currencyCode.append(contentsOf: results.rates.keys)
+                self.values.append(contentsOf: results.rates.values)
+                retreiveDataKeys(self.currencyCode)
+                retrieveDataValues(self.values)
+            } catch {
+                print(error)
+            }
+        }.resume()
+    }
+}
+
+final class CurrencyExchangeLocalData {
+    var currencyExchangeService: CurrencyExchangeConverterService?
+    
+    init(currencyExchangeService: CurrencyExchangeConverterService?) {
+        self.currencyExchangeService = currencyExchangeService
+    }
+    
+    func setDataToUserDefaults() {
+        currencyExchangeService?.fetchJSON(retreiveDataKeys: { keys in
+            UserDefaults.standard.set(keys, forKey: Keys.mostRecentExchangeRatesKeys)
+        }, retrieveDataValues: { values in
+            UserDefaults.standard.set(values, forKey: Keys.mostRecentExchangeRatesValues)
+        })
+    }
+    
+    func retrieveDataFromUserDefaults() {
+        let currencyKeys = UserDefaults.standard.stringArray(forKey: Keys.mostRecentExchangeRatesKeys)
+        let currencyValues = UserDefaults.standard.object(forKey: Keys.mostRecentExchangeRatesValues)
+        print(currencyKeys ?? "No such keys in User Defaults")
+        print(currencyValues ?? "No such values in User Defaults")
+    }
+}
+
+let currencyExchangeService = CurrencyExchangeConverterService()
+let currencyExchangeLocalData = CurrencyExchangeLocalData(currencyExchangeService: currencyExchangeService)
+currencyExchangeLocalData.setDataToUserDefaults()
+currencyExchangeLocalData.retrieveDataFromUserDefaults()
+
+///When we have an object from the backend, we have a possibility to create MVVM-C architecture and use data to display on UI part
+///This is example of MVVM-C structure of our project
+
+///Base coordinator conforms to different coordinators to use methods of our currency exchanges
+class BaseCoordinator {
+    private(set) weak var rootController: UINavigationController?
+    
+    init(rootController: UINavigationController) {
+        self.rootController = rootController
+    }
+    ///This is method which can be used in different coordinators to pushViewControllers, this method is not the only one which we can provide.
+    ///We can extend our base coordinator to handle different situatuations
+    func pushViewController(_ viewController: UIViewController) {
+        pushViewController(viewController)
+    }
+    ///This function needs to display that we can handle our dependencies in this class, but if we need some kind of parametres - we can create a new start function to avoid overriding this one.
+    func start() { assertionFailure("Must be overriden in subclass") }
```

Comment:
It's easy to forget to override in such cases, I would suggest to provide base coordinator  interface via a protocol